### PR TITLE
add offline mode support

### PR DESCRIPTION
**Pull request by DXist**
`--offline` flag allows to run local experimens without tracking in Polyaxon. Tested with a docker experiment.

Requires https://github.com/polyaxon/polyaxon-client/pull/30

### CODE REVIEW

File: polyaxon_cli/cli/run.py [hunk 1]
```diff
@@ -135,6 +137,9 @@ def run(ctx,
     tags = validate_tags(tags)
 
     if local:
+        if ctx.obj["offline"]:
+            os.environ['POLYAXON_IS_OFFLINE'] = 'true'
```
Comment (mouradmourafiq) [inline polyaxon_cli/cli/run.py]: Does this get reverted from the `environ` when the user does not pass the offline flag? If not then it's better no to wither:
 * pass the flag to other components that require it
 * rely entirely on the user's env var from the start

File: polyaxon_cli/run/docker.py [hunk 2]
```diff
@@ -124,7 +127,7 @@ def _create_docker_build(build_job, build_config, project):
             raise PolyaxonShouldExitError('')
 
         image_tag = hash_value(rendered_dockerfile)
-        if POLYAXON_NO_OP_KEY not in os.environ:
+        if not settings.IS_OFFLINE:
```
Comment (mouradmourafiq) [inline polyaxon_cli/run/docker.py]: I think it's better to use `if 'POLYAXON_IS_OFFLINE' in os.environ:` similar to [line](https://github.com/polyaxon/polyaxon-cli/pull/36/files#diff-4bc442d573f1f53e5ed495733f227038R48), without having to depend on the `settings` coming from the client.

File: polyaxon_cli/cli/run.py [hunk 3]
```diff
@@ -135,6 +137,9 @@ def run(ctx,
...
+            settings.IS_OFFLINE = True
```
Comment (mouradmourafiq) [inline polyaxon_cli/cli/run.py]: Is this required?

Comment (DXist) [reply polyaxon_cli/cli/run.py]: Polyaxon-client uses `settings.IS_OFFLINE` to check for offline mode.

settings monkeypatching is not required if we import `polyaxon_client.settings` after `POLYAXON_IS_OFFLINE` env variable is set.

Comment (mouradmourafiq) [reply polyaxon_cli/cli/run.py]: I understand, but I don't see why this change here, I would either do that on the main close to where we set the offline on the context, this way any command check the `offline` flag does not have to duplicate code, or not do it at all, as far as I can see, we only need to check for the offline, at the moment, in the local mode to add an export to the dockerfile, which we can do just with the env vars.

Comment (DXist) [reply polyaxon_cli/cli/run.py]: In local mode we need to create BuildJob and Experiment instances - they also need offline mode.

File: polyaxon_cli/run/docker.py (hunk 2 above)
Comment (DXist) [reply polyaxon_cli/run/docker.py]: Fixed this.

File: polyaxon_cli/cli/run.py (hunk 3 above)
Comment (DXist) [reply polyaxon_cli/cli/run.py]: I moved offline mode handling to `polyaxon_cli/main.py`

File: polyaxon_cli/cli/run.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 from __future__ import absolute_import, division, print_function
 
+import os
```
Comment (mouradmourafiq) [inline polyaxon_cli/cli/run.py]: Minor: unused import

File: polyaxon_cli/main.py [hunk 5]
```diff
@@ -57,7 +63,12 @@ def cli(context, verbose):
     non_check_cmds = [
         'completion', 'config', 'version', 'login', 'logout', 'deploy', 'admin', 'teardown'
     ]
-    if context.invoked_subcommand not in non_check_cmds:
+    context.obj = context.obj or {}
+    context.obj["offline"] = offline
+    if offline:
+        os.environ['POLYAXON_IS_OFFLINE'] = 'true'
```
Comment (mouradmourafiq) [inline polyaxon_cli/main.py]: Minor: can we use booleans instead, `'false'` would be evaluated to `True` as well

File: polyaxon_cli/main.py [hunk 6]
```diff
@@ -57,7 +63,12 @@ def cli(context, verbose):
...
+        settings.IS_OFFLINE = 'true'
```
Comment (mouradmourafiq) [inline polyaxon_cli/main.py]: Same as above.

File: polyaxon_cli/main.py (hunk 5 above)
Comment (mouradmourafiq) [reply polyaxon_cli/main.py]: Actually, you are right, and also the config manager can parse bool strings

File: polyaxon_cli/main.py (hunk 6 above)
Comment (mouradmourafiq) [reply polyaxon_cli/main.py]: May be this one only could be set to `True`.

**Comment by DXist** [comment]
@mouradmourafiq Fixed the boolean value and lint errors.

**Pull request merged**